NetKet/Optimizer/stochastic_reconfiguration.hpp

```diff
@@ -0,0 +1,162 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NETKET_SR_HPP
+#define NETKET_SR_HPP
+
+#include <complex>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
+#include <string>
+#include <vector>
+
+#include <Eigen/Dense>
+#include <Eigen/IterativeLinearSolvers>
+
+#include "Utils/parallel_utils.hpp"
+#include "Utils/random_utils.hpp"
+#include "common_types.hpp"
+#include "matrix_replacement.hpp"
+
+namespace netket {
+
+// Generalized Stochastic Reconfiguration Updates
+class SR {
+  double sr_diag_shift_;
+  bool use_iterative_;
+  bool use_cholesky_;
+  bool is_holomorphic_;
+
+  Eigen::MatrixXd Sreal_;
+  Eigen::MatrixXcd Scomplex_;
+
+ public:
+  SR() { setDefaultParameters(); }
+
+  void ComputeUpdate(const Eigen::Ref<const Eigen::MatrixXcd> Oks,
+                     const Eigen::Ref<const Eigen::VectorXcd> grad,
+                     Eigen::Ref<Eigen::VectorXcd> deltaP) {
+    double nsamp = Oks.rows();
+
+    SumOnNodes(nsamp);
+    auto npar = grad.size();
+    if (is_holomorphic_) {
+      if (!use_iterative_) {
+        // Explicit construction of the S matrix
+        Scomplex_.resize(npar, npar);
+        Scomplex_ = (Oks.adjoint() * Oks);
+        SumOnNodes(Scomplex_);
+        Scomplex_ /= nsamp;
+
+        // Adding diagonal shift
+        Scomplex_ += Eigen::MatrixXd::Identity(npar, npar) * sr_diag_shift_;
+
+        if (use_cholesky_ == false) {
+          Eigen::FullPivHouseholderQR<Eigen::MatrixXcd> qr(npar, npar);
+          qr.setThreshold(1.0e-6);
+          qr.compute(Scomplex_);
+          deltaP = qr.solve(grad);
+        } else {
+          Eigen::LLT<Eigen::MatrixXcd> llt(npar);
+          llt.compute(Scomplex_);
+          deltaP = llt.solve(grad);
+        }
+      } else {
+        Eigen::ConjugateGradient<SrMatrixComplex, Eigen::Lower | Eigen::Upper,
+                                 Eigen::IdentityPreconditioner>
+            it_solver;
+        // Eigen::GMRES<MatrixReplacement, Eigen::IdentityPreconditioner>
+        // it_solver;
+        it_solver.setTolerance(1.0e-3);
+        SrMatrixComplex S;
+        S.attachMatrix(Oks);
+        S.setShift(sr_diag_shift_);
+        S.setScale(1. / nsamp);
+
+        it_solver.compute(S);
+        deltaP = it_solver.solve(grad);
+        MPI_Barrier(MPI_COMM_WORLD);
+      }
+    } else {
+      if (!use_iterative_) {
+        // Explicit construction of the S matrix
```

Comment:
Not sure if it's actually wrong, but it does look weird to me... So if I'm reading this correctly, what you have here is `Sδw = F` where `S` and `F` are complex and `δw` should be real. Then, you decide to solve `Re[S]δw = Re[F]` instead of the initial system. I'm wondering why? I mean, obviously, it's very unlikely that `Sδw = F` will have a purely real solution. But then I'd expect one to resort to solving a LLS (Linear Least Square) problem: `min‖Sδw - F‖`.

Reply:
This is a bit subtle indeed. The system Re[S]dw=Re[F] is the correct way of doing the SR update when the function is not holomorphic. When the function is holomorphic instead you solve S*dw=F , and consider both real and imaginary part of dw... I have some notes on that, I will maybe attach them to the repository as technical notes

Reply:
To some extent this is explained also in this [book](https://www.amazon.com/Quantum-Monte-Approaches-Correlated-Systems/dp/1107129931), around page 139. But it's a bit obscure from that derivation why those formulas hold

Reply:
Please, do! Some time ago my numerical experiments showed the opposite so that's why I was asking...